File: LinkedList/40-combination-sum-ii/combination-sum-ii.cpp
class Solution {
public:
    void numFinder(vector<int>& candidates, vector<vector<int>>& ans, vector<int>& temp, int sum, int start, int target) {
        if (sum == target) {
            ans.push_back(temp);
            return;
        }

        for (int i = start; i < candidates.size(); ++i) {
            if (i > start && candidates[i] == candidates[i - 1]) {
                continue;  // Skip duplicates
            }

            if (sum + candidates[i] <= target) {
                temp.push_back(candidates[i]);
                sum = sum + candidates[i];
                numFinder(candidates, ans, temp,sum, i + 1, target);
                
                //backtracking ho raha guys 
                
                sum = sum - candidates[i];
                temp.pop_back();
            }
        }
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());  // Sort the candidates to handle duplicates.
        vector<vector<int>> ans;
        vector<int> temp;
        numFinder(candidates, ans, temp, 0, 0, target);
        return ans;
    }
};

File: DP/40-combination-sum-ii/combination-sum-ii.java
class Solution {

    private static void recur(int[] candidates,int currIndex,int target,List<Integer> curr,List<List<Integer>> ans) {

        if(target == 0 ) {
            ans.add(new ArrayList<>(curr));
            return ;
        }

        if(target < 0 || candidates.length == currIndex) {
            return ;
        }

        int i = currIndex ;

        curr.add(candidates[currIndex]);

        recur(candidates,currIndex + 1,target - candidates[currIndex],curr,ans);

       while(i < candidates.length && candidates[i] == candidates[currIndex]) {
            i++;
        }

        curr.remove(curr.size() - 1);

        recur(candidates,i,target,curr,ans);
    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {

        Arrays.sort(candidates);
        
        List<List<Integer>> ans = new ArrayList<>();

        List<Integer> curr = new ArrayList<>();

        recur(candidates,0,target,curr,ans);

        return ans ;
    }
}